crypto/params_dup.c

```diff
@@ -0,0 +1,185 @@
+            dst->data = buf[is_secure].cur;
+        }
+
+        if (in->data_type == OSSL_PARAM_OCTET_PTR
+            || in->data_type == OSSL_PARAM_UTF8_PTR) {
+            param_sz = sizeof(in->data);
+            if (has_dst)
+                *((const void **)dst->data) = *(const void **)in->data;
+        } else {
+            param_sz = in->data_size;
+            if (has_dst)
+                memcpy(dst->data, in->data, param_sz);
+        }
+        if (in->data_type == OSSL_PARAM_UTF8_STRING)
+            param_sz++; /* NULL terminator */
+        blks = ossl_param_bytes_to_blocks(param_sz);
+
+        if (has_dst) {
+            dst++;
+            buf[is_secure].cur += blks;
+        } else {
+            buf[is_secure].blocks += blks;
+        }
+        if (param_count != NULL)
+            *param_count += 1;
+    }
+    return dst;
+}
+
+OSSL_PARAM *OSSL_PARAM_dup(const OSSL_PARAM *src)
+{
+    size_t param_blocks;
+    OSSL_PARAM_BUF buf[OSSL_PARAM_BUF_MAX];
+    OSSL_PARAM *last, *dst;
+    int param_count = 1; /* Add terminator */
+
+    if (src == NULL)
+        return NULL;
+
+    memset(buf, 0, sizeof(buf));
+
+    /* First Pass: get the param_count and block sizes required */
+    (void)ossl_param_dup(src, NULL, buf, &param_count);
+
+    param_blocks = ossl_param_bytes_to_blocks(param_count * sizeof(*src));
+    /*
+     * The allocated buffer consists of an array of OSSL_PARAM followed by
+     * aligned data bytes that the array elements will point to.
+     */
+    if (!ossl_param_buf_alloc(&buf[OSSL_PARAM_BUF_PUBLIC], param_blocks, 0))
+        return NULL;
+
+    /* Allocate a secure memory buffer if required */
+    if (buf[OSSL_PARAM_BUF_SECURE].blocks > 0
+        && !ossl_param_buf_alloc(&buf[OSSL_PARAM_BUF_SECURE], 0, 1)) {
+        OPENSSL_free(buf[OSSL_PARAM_BUF_PUBLIC].alloc);
+        return NULL;
+    }
+
+    dst = (OSSL_PARAM *)buf[OSSL_PARAM_BUF_PUBLIC].alloc;
+    last = ossl_param_dup(src, dst, buf, NULL);
+    /* Store the allocated secure memory buffer in the last param block */
+    ossl_param_set_secure_block(last, buf[OSSL_PARAM_BUF_SECURE].alloc,
+                                buf[OSSL_PARAM_BUF_SECURE].alloc_sz);
+    return dst;
+}
+
+OSSL_PARAM *OSSL_PARAM_merge(const OSSL_PARAM *p1, const OSSL_PARAM *p2)
+{
+    const OSSL_PARAM *p;
+    OSSL_PARAM *params, *dst, *found;
+    int count = 1; /* Add terminator */
+
+    if (p1 == NULL)
+        return NULL;
+
+    for (p = p1; p->key != NULL; p++)
+        count++;
+    if (p2 != NULL) {
+        for (p = p2; p->key != NULL; p++)
+            count++;
+    }
+    params = OPENSSL_zalloc(count * sizeof(*p1));
+    if (params == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+    dst = params;
+    for (p = p1; p->key != NULL; p++)
+        *dst++ = *p;
+    if (p2 != NULL) {
+        for (p = p2; p->key != NULL; p++) {
+            found = OSSL_PARAM_locate(params, p->key);
```

Comment:
This is a quadratic time algorithm.  Sort the OSSL_PARAM arrays (using an extra array for indirection if necessary) and process both in linear time.

This can be done by sorting `params` and `p2` (indirectly) but it is better done by sorting both `p1` and `p2` (indirectly) up front and merging them.

Reply:
You mean a mergesort:)

Reply:
Kind of a mergesort, yes (the difference being that you are replacing identically valued items rather than including both).  It won't be slower for many cases.  Quadratic time is sucky and becomes significant very quickly.

The question then becomes: should the first or second item have precedence?

Reply:
The question is resolved in the .pod file..

Reply:
Ok updated, code looks way worse than the first pass.. Any suggestions are welcome.
I thought about using the openssl stack but this seems a bit too much.. It uses fixed arrays to avoid doing yet another pass over the data.